plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java

```diff
@@ -1001,6 +1012,171 @@ public void rollback()
+            }
+            else {
+                // Create a new column handle
+                IcebergColumnHandle baseColumnHandle = (IcebergColumnHandle) assignments.get(projectedColumn.getVariable().getName());
+                projectedColumnHandle = createProjectedColumnHandle(baseColumnHandle, projectedColumn.getDereferenceIndices(), expression.getType());
+                projectedColumnName = projectedColumnHandle.getQualifiedName();
+            }
+
+            Variable projectedColumnVariable = new Variable(projectedColumnName, expression.getType());
+            Assignment newAssignment = new Assignment(projectedColumnName, projectedColumnHandle, expression.getType());
+            newAssignments.put(projectedColumnName, newAssignment);
+
+            newVariablesBuilder.put(expression, projectedColumnVariable);
+            projectedColumnsBuilder.add(projectedColumnHandle);
+        }
+
+        // Modify projections to refer to new variables
+        Map<ConnectorExpression, Variable> newVariables = newVariablesBuilder.build();
+        List<ConnectorExpression> newProjections = projections.stream()
+                .map(expression -> replaceWithNewVariables(expression, newVariables))
+                .collect(toImmutableList());
+
+        List<Assignment> outputAssignments = newAssignments.build().values().stream().collect(toImmutableList());
+        IcebergTableHandle newTableHandle = new IcebergTableHandle(
+                icebergTableHandle.getSchemaName(),
+                icebergTableHandle.getTableName(),
+                icebergTableHandle.getTableType(),
+                icebergTableHandle.getSnapshotId(),
+                icebergTableHandle.getUnenforcedPredicate(),
+                icebergTableHandle.getEnforcedPredicate(),
+                Optional.of(projectedColumnsBuilder.build()));
+        return Optional.of(new ProjectionApplicationResult<>(
+                newTableHandle,
+                newProjections,
+                outputAssignments,
+                false));
+    }
+
+    /**
+     * Returns the assignment key corresponding to the column represented by {@param projectedColumn} in the {@param assignments}, if one exists.
+     * The variable in the {@param projectedColumn} can itself be a representation of another projected column. For example,
+     * say a projected column representation has variable "x" and a path=[0]. "x" can in-turn map to a projected
+     * column handle with base="a" and [1, 2] as a fieldId path. Then the method searches for a column handle in
+     * {@param assignments} with base="a" and path=[1, 2, 0].
+     * Note that the ProjectedColumnRepresentation uses index based dereferences but the IcebergColumnHandle#path uses Iceberg field ids.
+     */
+    public static Optional<String> find(Map<String, ColumnHandle> assignments, HiveApplyProjectionUtil.ProjectedColumnRepresentation projectedColumn)
```

Comment:
this seems like a bit too much aerobics for what it's worth. Can we create a columnhandle and compare that instead before adding if that's simpler?